Make one-shot stats faster

xinfengliu opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Kuberenets CRI interface `ListContainerStats` requires to collect every pod container's stats. Currently `daemon/stats/collector.go` runs in a serial way and sleep `1s` at the end of `for` loop. This causes `ListContainerStats` to suffer long delays if there are many pod containers on the node. 

This PR is to address this issue.

**- How I did it**

- Make the stats collector parallel with the number of CPUs.
- ~`Sleep` is not needed in a normal loop since a condition variable is used. Only sleep when an error happens at system level, then retry in next loop~. `time.Sleep(s.interval)` is needed for `one-short=0` or `stream=1`, otherwise the caller would receive successive stats without any interval.
- Move `bufReader` out of `Collector`, since it cannot be shared by multiple goroutines.
- Get container stats directly (i.e. without using publishers) for `one-shot` requests. 

**- How to verify it**

Run the PR build with `cri-dockerd`, observe the time taken for `ListContainerStats`. Without this code change, it could take over 3 seconds when about 40 pod containers on the node. ~With this fix, it takes less than 1 second~. 
- Making stats collection parallel reduces the `ListContainerStats` time to 2 seconds on the 4-CPU node. 
- Getting container stats directly for `one-shot` requests reduces the `ListContainerStats` time to less than 200ms. (The requests sent by `cri-dockerd` are `one-shot` requests)  

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
Improves the performance of the stats collector

**- A picture of a cute animal (not mandatory but encouraged)**

cpuguy83 commented:
https://github.com/Mirantis/cri-dockerd/pull/107 should have made stats collection very fast.
There is no sleeping (outside of a potential regression in moby or cri-dockerd) in this case.

xinfengliu commented:
`one-shot` works well when `len(s.publishers)==0`, because the registration of stats request can immediately wake up the collector. If the collector is busy with collecting stats or is sleeping after that, during this period of time, the new stats requests have to wait to be served in next loop, which means some delays. 

To avoid such delays, https://github.com/Mirantis/cri-dockerd/pull/38 set the concurrency for getting container stats to the number of containers,  if the number of containers is far larger than number of CPUs, It's not ideal.

corhere reviewed (changes requested):
Cache invalidation is famously one of the hard problems in computer science. This PR needs some more work on that front.

cpuguy83 commented:
It seems like the collector is really being used for 2 things:

1. Prevent multiple requests for stats for one container
2. As a giant mutex to only have one thing collecting stats at a time.

For 1, I think we can use a `singleflight` per container.
For 2, I'm not sure we should prevent this, but if we really want to we can just use a mutex so there is only one collection happening at a time.

xinfengliu commented on daemon/stats/collector.go:
```diff
@@ -20,7 +21,12 @@ type Collector struct {
 	interval   time.Duration
 	publishers map[*container.Container]*pubsub.Publisher
-	bufReader  *bufio.Reader
+	sync.RWMutex
```
Thanks for the reminder!

xinfengliu commented on daemon/stats/collector.go:
```diff
@@ -111,44 +121,88 @@ func (s *Collector) Run() {
+		// FIXME: move to containerd on Linux (not Windows)
+		stats.CPUStats.SystemUsage = systemUsage
+		s.RLock()
+		stats.CPUStats.OnlineCPUs = s.cache.onlineCPUs
```
Right, it's a problem. I wanted to cache it because `onlineCPUs` does not change often and I'm not sure if the cost of many concurrent `sched_getaffinity` syscalls on large systems.

xinfengliu commented on daemon/stats/collector.go:
```diff
@@ -111,44 +121,88 @@ func (s *Collector) Run() {
-					ID:   pair.container.ID,
-				})
-
-			default:
-				log.G(context.TODO()).Errorf("collecting stats for %s: %v", pair.container.ID, err)
-				pair.publisher.Publish(types.StatsJSON{
-					Name: pair.container.Name,
-					ID:   pair.container.ID,
-				})
 			}
+			return nil
+		})
+
+		numWorkers := int(onlineCPUs)
+		if numWorkers == 0 {
+			numWorkers = runtime.NumCPU()
+		}
+		if numWorkers > len(pairs) {
+			numWorkers = len(pairs)
 		}
 
+		for i := 0; i < numWorkers; i++ {
+			g.Go(func() error {
+				for pair := range pairCh {
+					stats, err := s.getContainerStats(pair.container)
+					if err != nil {
+						return err
+					}
+					pair.publisher.Publish(*stats)
+
+					select {
+					case <-ctx.Done():
+						return ctx.Err()
+					default:
+					}
+				}
+				return nil
+			})
+		}
+		g.Wait()
```
Thank you very much for sharing this valuable resource!

xinfengliu commented:
Not sure how to move forward.  The main purpose of this PR is to help `cri-dockerd`. If we can just take `one-shot` ContainerStats requests out of publishing channels, I believe it would be a help. Making collecting stats parallel in `s.Run()` loop does not bring much benefits based on my testing, `s.supervisor.GetContainerStats` is very fast, just a few `ms` usually. And `len(s.publishers)` would be small if we collect stats directly for `one-shot` requests.

corhere commented:
The stats API has three modes of operation, summarized by the following table:

| `Stream` | `OneShot` | Description |
|--------|--------|--------|
| True | True | (error) |
| True | False | Stream a container's stats sampled at ~1s intervals |
| False | True | Return a single sample of the container's stats |
| False | False | Return two samples of the container's stats, collected ~1s apart | 

For the streaming mode it seems reasonable to have a single collector which broadcasts results to all interested clients to amortize the cost of collection as the clients expect not to have control over the sampling frequency or phase. Similarly, clients requesting "two-shot" stats expect there to be some latency so the additional latency and jitter of using the collector to amortize the sampling costs is not a deal-breaker.

Then there are the clients that want more control over when a sample is taken. [Prometheus exporters](https://prometheus.io/docs/instrumenting/writing_exporters/#scheduling) for instance are expected to capture a sample only upon request so that the scraper has full control over the sampling interval and phase, timestamps the sample using its own clock to sidestep issues with clock skew, has visibility into the end-to-end latency, and can compensate to reduce jitter. For one-shot stats requests to be suitable for that use case, the stats sample needs to be taken synchronously with the request and not coalesced with other requests. Therefore I think what @xinfengliu has done in this PR to handle one-shot stats requests synchronously, without any single-flighting or queueing, is the right design choice.

-----

On the topic of caching `onlineCPUs`, I don't think any caching is necessary _if_ we change how that data is collected. The reporting of online CPUs was added in https://github.com/moby/moby/commit/115f91d7575d6de6c7781a96a082f144fd17e400 for reasons explained in the commit message. Initially `C.sysconf(C._SC_NPROCESSORS_ONLN)` was used to learn the number of online CPUs, but it was replaced in https://github.com/moby/moby/commit/cf104d85c35947c25dcd86cef19aa97fe31e4bbd with an implementation based on the `sched_getaffinity` syscall to remove the hard dependency on cgo. What's really interesting is how glibc implements `sysconf(_SC_NPROCESSORS_ONLN)`: it's just a wrapper around the GNU-specific [`get_nprocs(3)`](https://man7.org/linux/man-pages/man3/get_nprocs_conf.3.html) for POSIX compliance. [The implementation](https://github.com/bminor/glibc/blob/5d00c201b9a2da768a79ea8d5311f257871c0b43/sysdeps/unix/sysv/linux/getsysstats.c#L213-L222) is very informative.
1. First it tries to read and parse `/sys/devices/system/cpu/online`
2. If that fails, it tries to read and parse `/proc/stat`
3. And if that also fails, it falls back to `sched_getaffinity`

And wouldn't you know, `func (*Collector) getSystemCPUUsage()` already reads `/proc/stat`, _and it is called in the inner loop!_ [The kernel only writes `cpu%d` lines for online CPUs](https://elixir.bootlin.com/linux/v6.5.4/source/fs/proc/stat.c#L139) and we're already paying the cost of reading the file each time we sample a container's stats, so it would not require any additional syscalls (read: be practically free) to also determine the number of online cpus by counting the number of `cpu%d` lines.

tl;dr modify `getSystemCPUUsage()` to also return the number of online CPUs so no caching is needed for performance.

xinfengliu commented:
@corhere 

Thank you very much for the deep analysis. I will modify the PR per your suggestions. 

As for collecting stats inside `Collector.Run()`, I feel it does not need to be parallel, I'd like to revert the related codes. Do  you have opinions on this?

corhere commented:
> As for collecting stats inside `Collector.Run()`, I feel it does not need to be parallel, I'd like to revert the related codes. Do you have opinions on this?

Revert away!

xinfengliu committed:
Make one-shot stats faster

This commit moves one-shot stats processing out of the publishing
channels, i.e. collect stats directly.

Also changes the method of getSystemCPUUsage() on Linux to return
number of online CPUs also.

Signed-off-by: Xinfeng Liu <XinfengLiu@icloud.com>

xinfengliu commented:
@corhere 
I have made the code modifications based on the feedbacks and changed the PR title. Please take a look. Thanks.

corhere reviewed (commented):
I see no logic bugs in the code. There is still some room for further improvement, however.

xinfengliu commented:
@corhere 

Thanks for your suggestions. I have made the modifications.

As for `daemon.GetContainerStats(ctr)`, I changed a little bit to the logic: 

- for `errdefs.ErrConflict` and `errdefs.ErrNotFound` (e.g. container not found or not running), return an empty `stats` and a  `nil` error. This is consistent to the previous user experiences. 

-  for other errors (which usually means errors not caused by the user), return the `err`. In this case, `Collector.Run` publishes an empty `stats`. The only difference is: previously if `getSystemCPUUsage` returns err, `Collector.Run` does not publish the `stats`, now in any case `Collector.Run` publishes the `stats`.

BTW, the currently [failed CI job](https://github.com/moby/moby/actions/runs/6322097037/job/17167203121?pr=46448) seems not related to my PR and I don't have permission to re-run only failed CI jobs.

corhere reviewed (approved):
LGTM! I'm not super thrilled with where errors are handled and logged, but it is consistent with existing behaviour and the code is already way cleaner than before.

xinfengliu commented on daemon/stats.go:
```diff
@@ -146,17 +156,37 @@ func (daemon *Daemon) unsubscribeToContainerStats(c *container.Container, ch cha
+
+	switch err.(type) {
+	case nil:
+		// Sample system CPU usage close to container usage to avoid
+		// noise in metric calculations.
+		systemUsage, onlineCPUs, err := getSystemCPUUsage()
```
@corhere I'm not clear about this. Could you suggest the order?

corhere replied:
Container usage (1) is sampled with `daemon.stats(container)`. System CPU usage (3) is sampled with `getSystemCPUUsage()`.  In between those calls is a call to `daemon.getNetworkStats(container)`. "Sample system CPU usage close to container usage to avoid noise in metric calculations."

xinfengliu replied:
@corhere I pushed a new [commit](https://github.com/moby/moby/commit/3d70af4ee46874fbc859878f3f22ee4f6aa5f5f3) today for this minor change. Please check if it is what you want. Thanks.

thaJeztah commented on daemon/stats.go:
```diff
@@ -43,6 +44,15 @@ func (daemon *Daemon) ContainerStats(ctx context.Context, prefixOrName string, c
 
+	// Get container stats directly if OneShot is set
+	if config.OneShot {
+		stats, err := daemon.GetContainerStats(ctr)
```
@corhere @cpuguy83 I saw an earlier mention of using singleflight; https://github.com/moby/moby/pull/46448#issuecomment-1724104569

> For 1, I think we can use a singleflight per container.

Would this be the path for which to consider a singleflight, or would that be a premature optimisation?

corhere replied:
I rebutted against using singleflight in https://github.com/moby/moby/pull/46448#issuecomment-1731942882

thaJeztah replied:
oh! I saw the part about caching, but missed this part;

> ... and not coalesced with other requests.

all good 👍

xinfengliu committed:
adjust stats sampling order

Signed-off-by: Xinfeng Liu <XinfengLiu@icloud.com>

xinfengliu commented:
@thaJeztah 
Can we port this PR to `23.0` ? If yes, what do I need to do? Thanks.

thaJeztah commented:
@xinfengliu let me defer that question to @cpuguy83 and @corhere who had a much closer look into these changes.

FWIW; we'll be working on 25.0 releases soon as well; so we'll have to look which branches we want to backport to (if any); was there a specific reason you needed this in the 23.0 branch?

corhere commented:
@xinfengliu open a backport PR targeting the 23.0 branch which cherry-picks these commits. (`git cherry-pick -sx`)

xinfengliu commented:
The backport PR for 23.0 branch #46617 has been created and CI has passed. 

@thaJeztah 
Most our customers just started to upgrade to MCR 23.0.x and they might stay on 23.0.x for a long time, so the backport will be beneficial to them. (If you want me to backport it to moby 24.0 branch as well, I can create another PR.)

thaJeztah commented:
@xinfengliu can you also open a backport for the 24.0 branch, otherwise 24.0 would "regress" compared to 23.0

xinfengliu commented:
@thaJeztah 
Sure. the backport for 24.0 branch has been created. https://github.com/moby/moby/pull/46619
